app2/ota_metadata.py

```diff
@@ -129,70 +165,47 @@ def _parse_metadata(self, metadata_jwt):
         # parse metadata.jwt payload
         return self._parse_payload(self._payload_json)
 
-    def verify(self, certificate_pem):
-        """"""
-        try:
-            certificate = crypto.load_certificate(crypto.FILETYPE_PEM, certificate_pem)
-            logger.debug(f"certificate: {certificate}")
-            verify_data = self.get_header_payload().encode()
-            logger.debug(f"verify data: {verify_data}")
-            crypto.verify(certificate, self._signature, verify_data, "sha256")
-            return True
-        except Exception as e:
-            logger.exception("Verify error:")
-            return False
-
-    def get_signature(self):
-        """
-        Get signature
-        """
-        return self._signature
-
-    def get_header_payload(self):
+    def _get_header_payload(self):
         """
         Get Header and Payload urlsafe base64 string
         """
         jwt_list = self.__metadata_jwt.split(".")
         return jwt_list[0] + "." + jwt_list[1]
 
-    def get_directories_info(self):
-        """
-        return
-            directory file info list: { "file": file name, "hash": file hash }
-        """
-        return self.__metadata_dict["directory"]
-
-    def get_symboliclinks_info(self):
-        """
-        return
-            symboliclink file info: { "file": file name, "hash": file hash }
-        """
-        return self.__metadata_dict["symboliclink"]
-
-    def get_regulars_info(self):
-        """
-        return
-            regular file info: { "file": file name, "hash": file hash }
-        """
-        return self.__metadata_dict["regular"]
-
-    def get_persistent_info(self):
-        """
-        return
-            persistent file info: { "file": file name, "hash": file hash }
-        """
-        return self.__metadata_dict["persistent"]
-
-    def get_rootfsdir_info(self):
-        """
-        return
-            rootfs_directory info: {"file": dir name }
-        """
-        return self.__metadata_dict["rootfs_directory"]
-
-    def get_certificate_info(self):
-        """
-        return
-            certificate file info: { "file": file name, "hash": file hash }
-        """
-        return self.__metadata_dict["certificate"]
+    def _verify_certificate(self, certificate: str):
+        certs = [cert.name for cert in sorted(list(self._certs_dir.glob("*.*.pem")))]
```

Comment:
I think we can make the root CA and intermediate CA into one bundle, a.k.a., combine these CAs together into one file, passing the whole certificate chain bundle and let the pyOpenssl to handle the order of the certificates?

I am writing a small scale codes to check if it is the case.

Reply:
> combine these CAs together into one file

Yes, we can do it and it is in a standard format. But I just didn't want to write the split code.

Reply:
I have done a small scale test.

I prepare 1 cert for verification, and 3 chains of certificates under the working folder, and the test code is as follow:
```
def verify(certs_dir: Path, cert_to_verify):
    certs_list = [c.name for c in certs_dir.glob("*.crt")]
    load_pem = partial(crypto.load_certificate, crypto.FILETYPE_PEM)

    store = crypto.X509Store()
    for c in certs_list:
        store.add_cert(load_pem(open(c).read()))
    
    cert_loaded = load_pem(open(cert_to_verify).read())
    try:
        store_ctx = crypto.X509StoreContext(store, cert_loaded)
        store_ctx.verify_certificate()
    except crypto.X509StoreContextError as e:
        print("verification failed: ", e)
```
The code works as expected.
It is totally OK to load all certificates at once, and the **OpenSSL.crypto.X509Store** can handle them and automatically verify the certificate with the right CA chain.

Reply:
EDIT: this code cannot handling the situation when all the CA chains are sharing the same Common name.